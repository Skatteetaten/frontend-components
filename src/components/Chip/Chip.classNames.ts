import { getFocusStyle } from '..';
import { getTheme } from '@uifabric/styling';
import { mergeStyles } from '@uifabric/merge-styles';
import { PaletteProps } from '..';
// @ts-ignore TODO
export var getClassNames = function getClassNames(props) {
  const palette = getTheme().palette as PaletteProps;
  const inset = '-0.5rem';
  const radius = '3rem';

  return mergeStyles([
    // @ts-ignore TODO
    getFocusStyle({ palette }, inset, 'relative', radius),
    {
      displayName: `ske-chip`,
      display: 'inline-block',
      color: palette.skeColor.blackAlt,
      fontWeight: props.bold ? '700' : '400',
      border: `1px solid ${palette.skeColor.white}`,
      verticalAlign: 'middle',
      padding: '3px 12px',
      fontSize: props.size,
      borderRadius: '5px',
      // @ts-ignore TODO
      backgroundColor: palette.skeColor[props.type],
      selectors: {
        '& > div': {
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center'
        },
        '& a': {
          textDecoration: 'underlined',
          color: palette.skeColor.blackAlt
        },
        '& i': {
          verticalAlign: 'text-bottom'
        }
      }
    }
  ]);
};
