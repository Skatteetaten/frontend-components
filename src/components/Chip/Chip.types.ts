import * as React from 'react';

export enum ChipType {
  WARNING = 'lightPink',
  OK = 'lightGreen',
  NEUTRAL = 'beige',
}

export interface ChipProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Bruksområde som avgir hvilken farge chip-en får */
  type?: ChipType;
  /** Størrelse på Chip */
  size?: 'standard' | 'large';
  /** aria-label */
  ariaLabel?: string;
  /** Overstyring av stiler */
  className?: string;
}
