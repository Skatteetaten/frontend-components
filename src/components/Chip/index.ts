export { Chip as default } from './Chip';
export * from './Chip';
