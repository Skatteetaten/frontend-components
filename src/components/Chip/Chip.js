import React from 'react';
import PropTypes from 'prop-types';
import { getClassNames } from './Chip.classNames';

/**
 * @visibleName Chip (Merkelapp)
 */
export default class Chip extends React.PureComponent {
  static WARNING = 'lightPink';
  static OK = 'lightGreen';
  static NEUTRAL = 'beige';

  static propTypes = {
    /** Bruksområde som avgir hvilken farge chip-en får */
    type: PropTypes.oneOf([Chip.NEUTRAL, Chip.OK, Chip.WARNING]),
    /** Størrelse på Chip */
    size: PropTypes.oneOf(['standard', 'large']),
    /** Angir aria-label til skjermleser */
    ariaLabel: PropTypes.string
  };

  static defaultProps = {
    type: Chip.NEUTRAL,
    size: 'standard',
    ariaLabel: undefined
  };

  render() {
    const { ariaLabel, children, ...props } = this.props;
    return (
      <div
        aria-label={ariaLabel}
        className={getClassNames(this.props)}
        {...props}
      >
        {children}
      </div>
    );
  }
}
