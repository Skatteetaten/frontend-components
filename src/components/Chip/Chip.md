** Chip brukes til å markere eller utheve noe - som en merkelapp. Et typisk tilfelle er dersom man har en liste med opplysninger, og ett at elementene i listen er et unntak som brukeren bør raskt få øye på. Chips kan også brukes for å vise markere aktive filtre i et søk eller kategorier i artikler.**

Variasjon i utheving, størrelse og farge:

```js
import Chip from '@skatteetaten/frontend-components/Chip';
import Icon from '@skatteetaten/frontend-components/Icon';

<div>
  <Chip ariaLabel="Klagesak">Klagesak </Chip>
  <span> </span>
  <Chip type={Chip.OK} size="standard" ariaLabel="Godkjent">
    <Icon iconName="Check" /> Godkjent
  </Chip>
  <span> </span>
  <Chip type={Chip.WARNING} ariaLabel="Unntatt offentlighet">
    Unntatt offentlighet
  </Chip>
</div>;
```

Eksempel på bruk i tekst:

```js
import Chip from '@skatteetaten/frontend-components/Chip';

<div>
  <h3>Dekning av saksomkostninger</h3>
  <p>
    Klage på avslag om dekning av saksomkostninger etter lov av 13. juni 1980
    nr. 24 om ligningsforvalgning (likningsloven) §9-11 etter at skattekontoret
    omgjorde endringsvedtak for 2005 om internprising og tilleggsskatt.
  </p>
  <Chip ariaLabel="Tilleggsskatt">Tilleggsskatt</Chip> <span> </span>
  <Chip ariaLabel="Internprising">Internprising</Chip> <span> </span>
  <Chip ariaLabel="Skatteklagenemnda">Skatteklagenemnda</Chip> <span> </span>
</div>;
```

```js noeditor
import Accordion from '@skatteetaten/frontend-components/Accordion';
import AccordionItem from '@skatteetaten/frontend-components/Accordion/AccordionItem';

<Accordion>
  <AccordionItem
    isOpen
    toggleContent
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      Det er viktig å bruke dette med varsomhet slik at ikke informasjonen
      drukner i merkelapper.
    </p>
    <p>
      En chip er liten i sin natur, men dersom du ønsker et kraftigere visuelt
      uttrykk finnes den i en stor variant.
    </p>
    <p>
      Standard farge på chip er beige/gul, men vi har et par varianter for i
      grønn og rød dersom du trenger å skille chips fra hverandre. Bruk rød til
      obs/advarsel og grønn for å signalisere ok.
    </p>
    <p>
      Chips kan inneholde lenker dersom de er klikkbare. Lenkene skal være blå.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <ul>
      <li>
        Pass på at fargen og teksten på merkelappen alltid samsvarer med
        hverandre. Da forhindrer du at brukere med dårlig fargesyn, eller en
        dårlig skjerm, går glipp av informasjon.
      </li>
      <li>
        Bruk <b>aria-label</b> attributten slik at skjermlesere kan fange opp
        teksten.
      </li>
    </ul>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      Dette er en komponent vi har laget selv fra bunnen av, og det finnes
      derfor ikke flere tilgjengelige props.
    </p>
  </AccordionItem>
</Accordion>;
```
