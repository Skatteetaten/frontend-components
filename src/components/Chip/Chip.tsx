import * as React from 'react';
import { getClassNames } from './Chip.classNames';

enum ChipType {
  WARNING = 'lightPink',
  OK = 'lightGreen',
  NEUTRAL = 'beige'
}

interface ChipProps {
  /** Bruksområde som avgir hvilken farge chip-en får */
  type?: ChipType;
  /** Størrelse på Chip */
  size?: 'standard' | 'large';
}
/**
 * @visibleName Chip (Merkelapp)
 */
export default class Chip extends React.PureComponent<ChipProps, {}> {
  static WARNING = ChipType.WARNING;
  static OK = ChipType.OK;
  static NEUTRAL = ChipType.NEUTRAL;
  static defaultProps = {
    size: 'standard',
    type: Chip.NEUTRAL
  };

  render() {
    const { children, ...props } = this.props;
    return (
      <div className={getClassNames(this.props)} {...props}>
        {children}
      </div>
    );
  }
}
