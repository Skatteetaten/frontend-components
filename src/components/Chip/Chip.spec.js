import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Chip from './Chip';

function oppsettShallow(props) {
  return shallow(<Chip {...props} />);
}

function oppsettFullDOM(props) {
  return mount(<Chip {...props} />);
}

describe('Chip komponent', () => {
  it('matcher snapshot', () => {
    const wrapper = oppsettFullDOM();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('rendrer Chip med default props', () => {
    const wrapper = oppsettShallow();

    expect(wrapper.prop('type')).toEqual('beige');
    expect(wrapper.prop('aria-label')).toEqual(undefined);
    expect(wrapper.prop('size')).toEqual('standard');
  });

  it('setter CheckBox med riktige props', () => {
    const wrapper = oppsettShallow({
      size: 'large',
      type: Chip.OK
    });

    expect(wrapper.prop('type')).toEqual('lightGreen');
    expect(wrapper.prop('size')).toEqual('large');
  });
});
