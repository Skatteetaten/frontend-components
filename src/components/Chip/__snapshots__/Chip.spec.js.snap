// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip komponent matcher snapshot 1`] = `
<Chip
  size="standard"
  type="beige"
>
  <div
    className=

        {
          background-color: #f9ede2;
          border-radius: 5px;
          border: 1px solid #ffffff;
          color: #1a1a1a;
          display: inline-block;
          font-size: standard;
          font-weight: 400;
          padding-bottom: 3px;
          padding-left: 12px;
          padding-right: 12px;
          padding-top: 3px;
          vertical-align: middle;
        }
        & > div {
          align-items: center;
          display: flex;
          justify-content: center;
        }
        & a {
          color: #1a1a1a;
          text-decoration: underlined;
        }
        & i {
          vertical-align: text-bottom;
        }
    size="standard"
    type="beige"
  />
</Chip>
`;
